### PR TITLE
iqn validation

### PR DESCRIPTION
**Pull request by ayalash**

### CODE REVIEW

File: infi/dtypes/iqn/__init__.py [hunk 1]
```diff
@@ -30,13 +15,34 @@ def __hash__(self):
...
-        return not (self == other)
+        return not (self == other)  # pylint: disable=superfluous-parens
+
+
+class InvalidIQN(ValueError):
+    def __init__(self, name):
+        super(InvalidIQN, self).__init__('Invalid IQN {}'.format(name))
+
+
+
```
Comment (spatz) [inline infi/dtypes/iqn/__init__.py]: Too many blank lines, use 2 please.

File: infi/dtypes/iqn/__init__.py [hunk 2]
```diff
@@ -30,13 +15,34 @@ def __hash__(self):
...
                 return False
-        return self._name == other._name
+        return self._name == other._name  # pylint: disable=protected-access
 
     def __ne__(self, other):
-        return not (self == other)
+        return not (self == other)  # pylint: disable=superfluous-parens
+
+
+class InvalidIQN(ValueError):
```
Comment (spatz) [inline infi/dtypes/iqn/__init__.py]: Exception should be first in file.

File: infi/dtypes/iqn/__init__.py [hunk 3]
```diff
@@ -49,3 +55,10 @@ def get_extra(self):
 
     def get_extra_fields(self):
         return self._extra
+
+
+def iscsi_name_factory(iscsi_name):
```
Comment (spatz) [inline infi/dtypes/iqn/__init__.py]: Should be called `make_iscsi_name`.

File: tests/test_iqn.py [hunk 4]
```diff
@@ -18,6 +19,24 @@ def test_rfc_examples():
     assert a.get_extra() == ''
     assert a.get_extra_fields() == ()
 
-def test_copy_ctor():
-    a = IQN('iqn.2001-04.com.example')
-    assert IQN(a) == a
+@pytest.mark.parametrize('copy_types', [(IQN, IQN), (IQN, iSCSIName), (iSCSIName, IQN), (iSCSIName, iSCSIName)])
```
Comment (spatz) [inline tests/test_iqn.py]: Use `itertools.product`.

File: tests/test_iqn.py [hunk 5]
```diff
@@ -18,6 +19,24 @@ def test_rfc_examples():
...
+    source_type, target_type = copy_types
+    obj = source_type('iqn.2001-04.com.example')
+    assert target_type(obj) == obj
+
+def test_invalid_iqn():
+    iscsi_identifier = '1'
+    with pytest.raises(InvalidIQN):
+        IQN(iscsi_identifier)
+    obj = iSCSIName(iscsi_identifier)
+    assert iscsi_name_factory(iscsi_identifier) == obj
```
Comment (spatz) [inline tests/test_iqn.py]: Operator `==` is not enough, please check with `isinstance()`.

**Review by drrlvn — CHANGES_REQUESTED** [review]

File: infi/dtypes/iqn/__init__.py [hunk 6]
```diff
@@ -30,13 +15,34 @@ def __hash__(self):
     def __eq__(self, other):
         if not isinstance(other, IQN):
```
Comment (spatz) [inline infi/dtypes/iqn/__init__.py]: Should not be `IQN`.

**Review by drrlvn — APPROVED** [review]

**Pull request merged**